keps/0023-enable-disable-features.md

```diff
@@ -0,0 +1,131 @@
+---
+kep-number: 23
+title: Enabling and Disabling features in KUDO Operators
+authors:
+  - "@zmalik"
+owners:
+  - "@zmalik"
+creation-date: 2020-01-28
+last-updated: 2020-01-28
+status: provisional
+---
+
+# Enabling and Disabling features in KUDO Operators
+
+## Summary 
+
+KUDO enables Operator developers to create Operators and develop features for them. 
+The goal of this KEP is to facilitate operator developers to enable and disable the features in a deterministic way.
+
+## Motivation
+
+We want to provide operator users a way to enable and disable features in their operator. And right now there is no operator-user friendly or operator-developer friendly way to do it. 
+
+### Goals
+
+Given a feature, `A` is composed of Kubernetes object `x`, `y` and `z`
+ 
+- create or update kubernetes objects  `x`, `y` and `z`, when feature `A` is enabled.
+- delete kubernetes objects  `x`, `y` and `z`, when feature `A` is disabled.
+
+The goal of this KEP is to provide a way to create and delete objects `x`, `y` and `z`. 
+
+### Non-Goals
+
+- complex logic like loops to trigger multiple features
+- enable/disable nested features
+- evaluate multiple conditions to enable one feature
+
+## Proposal
+
+### Using Instance Parameters
+
+In KUDO, specific plans can be triggered when updating a parameter. Plans are specifically formed by phases, steps and utimately by tasks.
+And when talking about enabling/disabling features, we are eventually talking about the `Kind` of the Task. Currently, we have `Apply`, `Delete` and `Pipe` tasks kind.
+Enabling a feature is running a task of `Apply` kind and disabling is running a task of `Delete`
+
+We propose introducing a new Task kind named `Feature`. The Feature kind handles Apply or Delete resources based on the value of a parameter in the Operator Instance.
+
+A `Feature` task will look like:
+
+```
+tasks:
+  - name: mirror-maker-configuration
+    kind: Feature
+    spec:
+      parameter: MIRROR_MAKER_ENABLED
+      resources:
+        - mirror-maker-cm.yaml
+  - name: mirror-maker
```

Comment:
could this ultimately be one task with two resources? 🤔

Reply:
yes, but as we are moving towards making more sense of steps, I wanted to keep a `configuration` step and another `app` step 
but to answer your question yes, it could be only one task and one step for this example

Reply:
I think, for the purpose of understanding this KEP, one task (with two resources) would be sufficient.

Reply:
+1 to _one-word-kinds_. and my _heavy-plus_ goes to `kind: Toggle`, see my [reasoning](https://github.com/kudobuilder/kudo/pull/1317#pullrequestreview-350131081) for it.